lib/puppetserver/ca/create_action.rb

```diff
@@ -0,0 +1,268 @@
+        if config_path
+          errors = FileUtilities.validate_file_paths(config_path)
+          return 1 if Utils.handle_errors(@logger, errors)
+        end
+
+        # Load, resolve, and validate puppet config settings
+        puppet = PuppetConfig.parse(config_path)
+        return 1 if Utils.handle_errors(@logger, puppet.errors)
+
+        # Load most secure signing digest we can for csr signing.
+        signer = SigningDigest.new
+        return 1 if Utils.handle_errors(@logger, signer.errors)
+
+        # Make sure we have all the directories where we will be writing files
+        FileUtilities.ensure_dir(puppet.settings[:certdir])
+        FileUtilities.ensure_dir(puppet.settings[:privatekeydir])
+        FileUtilities.ensure_dir(puppet.settings[:publickeydir])
+
+        # Generate and save certs and associated keys
+        all_passed = generate_certs(certnames, puppet.settings, signer.digest)
+        return all_passed ? 0 : 1
+      end
+
+      # Create crls and keys, then submit them to CA, request for the CA to sign
+      # them, download the signed certificates from the CA, and finally save
+      # the signed certs and associated keys. Returns true if all certs were
+      # successfully created and saved.
+      def generate_certs(certnames, settings, digest)
+        passed = certnames.map do |certname|
+          key, csr = generate_key_csr(certname, settings, digest)
+          return false unless submit_certificate_request(certname, csr.to_s, settings)
+          return false unless sign_cert(certname, settings)
+          if download_cert(certname, settings)
+            save_keys(key, certname, settings)
+            true
+          else
+            false
+          end
+        end
+        passed.all?
+      end
+
+      def generate_key_csr(certname, settings, digest)
+        host = Puppetserver::Ca::Host.new(digest)
+        private_key = host.create_private_key(settings[:keylength])
+        csr = host.create_csr(certname, private_key)
+
+        return private_key, csr
+      end
+
+      def http_client(settings)
+        @client ||= RefactoredClient.new(settings[:localcacert],
+                                         settings[:certificate_revocation],
+                                         settings[:hostcrl])
+      end
+
+      # Make an HTTP request to submit certificate requests to CA
+      # @param certname [Array] the name of the certificate to fetch
+      # @param csr [String] string version of a OpenSSL::X509::Request
+      # @param settings [Hash] a hash of config settings
+      # @return [Boolean] success of all csrs being submitted to CA
+      def submit_certificate_request(certname, csr, settings)
+        client = http_client(settings)
+        url = client.make_ca_url(settings[:ca_server],
+                       settings[:ca_port],
+                       'certificate_request')
+
+        client.with_connection(url) do |connection|
+          url.resource_name = certname
```

Comment:
We should probably update the calls to `make_ca_url` to include the certname and then when we refactor out the CA update it to cache the connection.